### PR TITLE
Taxing of lines without tax class (SHOOP-2077)

### PR DESCRIPTION
**Pull request by suutari-ai**
Implement taxing of lines without a tax class, especially discount
lines.

Has some other related changes too.  See the commit messages for
details.

### CODE REVIEW

File: shoop_tests/functional/test_discount_taxing.py [hunk 1]
```diff
@@ -0,0 +1,544 @@
+from collections import defaultdict
```
Comment (tulimaki) [inline shoop_tests/functional/test_discount_taxing.py]: Add ``from __future__ import unicode_literals`` for python 2 also you might want to add the license header.

File: shoop/core/taxing/utils.py [hunk 2]
```diff
@@ -6,13 +6,53 @@
...
+    """
+    Generate tax class proportions from taxed lines.
+
+    Sum prices per tax class and return a list of (tax_class, factor)
+    pairs, where factor is the proportion of total price of lines with
+    given tax class from the total price of all lines.
+
+    :type lines: list[shoop.core.order_creator.SourceLine]
+    :param lines: List of taxed lines to generate proportions from
+
```
Comment (tulimaki) [inline shoop/core/taxing/utils.py]: Extra empty line

File: shoop_tests/functional/test_discount_taxing.py [hunk 3]
```diff
@@ -0,0 +1,544 @@
...
+        #    Name    Rate    Base amount     Tax amount         Taxful
+        assert get_pretty_tax_summary(source) == [
+            'Tax-A   0.25  144.000000000   36.000000000  180.000000000']
+        assert get_pretty_line_taxes_of_discount_lines(source) == [[
+            'Tax-A   0.25  -16.000000000   -4.000000000  -20.000000000']]
+    else:
+        assert get_pretty_tax_summary(source) == [
+            'Tax-A   0.25  180.000000000   45.000000000  225.000000000']
+        assert get_pretty_line_taxes_of_discount_lines(source) == [[
+            'Tax-A   0.25  -20.000000000   -5.000000000  -25.000000000']]
```
Comment (tulimaki) [inline shoop_tests/functional/test_discount_taxing.py]: Based on this. Should the line total be -20. If merchant use taxless prices in shop, is the only way to set discount amount 20 to actually calculate possible taxes. Sould the discount amount in campaings follow shop prices whether is taxless or taxfull.

File: shoop/core/taxing/utils.py (hunk 2 above)
Comment (suutari-ai) [inline shoop/core/taxing/utils.py]: Actually it was intentional, to separate input params from return param, but it seems that other docstrings don't have a line break there, so I'll remove this too. :)

File: shoop_tests/functional/test_discount_taxing.py (hunk 1 above)
Comment (suutari-ai) [inline shoop_tests/functional/test_discount_taxing.py]: Will do. Thanks.

**Comment by suutari-ai** [comment]
Fixed the unicode issue in the test module and added the license header. Removed the extra blank from docstring of `get_tax_class_proportions`. Also rebased to current master, since had conflicts in changelog.

**Comment by suutari-ai** [comment]
retest this please

**Pull request merged**